### ISSUE TITLE
Excluding contour triangles

### ISSUE THREAD

**Issue by iominh** [brendankenny/libtess.js#13]
Hi Brendan,

Thanks so much for creating this awesome library! I've learned a lot from using it. I've tried other JS libraries like earcut, poly2tri, and several point-in-polygon libraries, but libtess.js gives the best results across many different inputs.

This is a long shot and I'm not sure where to post this but maybe you might have some insights (or alternatively this could be rubber duck debugging for me). Anyway, I'm using a mapping library called CesiumJS to render arbitrary polygons on a globe. However, I want to exclude certain polygons from being rendered, so this is where libtess.js comes in.

Using libtess.js, I've implemented code very similarly to the [simple triangulation example you've provided](https://github.com/brendankenny/libtess.js/tree/gh-pages/examples/simple_triangulation), so initially I created a single contour of 2 triangles, as seen here (the blue and brown triangles):

![screen shot 2015-06-09 at 11 29 59 am](https://cloud.githubusercontent.com/assets/1434477/8062339/24baad9a-0e9d-11e5-883b-ea136a6bb482.png)

Next, I added another contour for an arbitrary polygon I want to not render (it's actually a "circle" polygon with 100 points that are radius-distanced from a center point). This produced the following results, where I render every triangle returned from the array passed into gluTessBeginPolygon: 

![screen shot 2015-06-09 at 11 30 46 am](https://cloud.githubusercontent.com/assets/1434477/8062403/7a6854a4-0e9d-11e5-9fab-69a0cb5ae960.png)

This worked better than every other library/algorithm I've tried (many others had issues triangulating the contours) but my goal is to exclude the center circle.

I tried setting the winding rule to every possible rule to no avail. However, the "GLU_TESS_WINDING_ABS_GEQ_TWO" rule seems to produce the interior contour/polygon that I'm trying to exclude, so that seemed interesting:

![glu_tess_winding_abs_geq_two](https://cloud.githubusercontent.com/assets/1434477/8062472/e2eb281c-0e9d-11e5-835e-832cfa522ff0.jpg)

Here are the other rules I tried

``` js
tessy.gluTessProperty(libtess.gluEnum.GLU_TESS_WINDING_RULE, libtess.windingRule.GLU_TESS_WINDING_NEGATIVE);
tessy.gluTessProperty(libtess.gluEnum.GLU_TESS_WINDING_RULE, libtess.windingRule.GLU_TESS_WINDING_ODD);
tessy.gluTessProperty(libtess.gluEnum.GLU_TESS_WINDING_RULE, libtess.windingRule.GLU_TESS_WINDING_NONZERO);
```

GLU_TESS_WINDING_NEGATIVE:
![glu_tess_winding_negative2](https://cloud.githubusercontent.com/assets/1434477/8065256/568a82bc-0eae-11e5-85a0-8aac7b9d02e1.png)

GLU_TESS_WINDING_POSITIVE:
![glu_tess_winding_positive](https://cloud.githubusercontent.com/assets/1434477/8065188/eb96a0f8-0ead-11e5-8012-eb20d960336b.png)

GLU_TESS_WINDING_NONZERO:
![glu_tess_winding_nonzero](https://cloud.githubusercontent.com/assets/1434477/8065203/fb9c705e-0ead-11e5-9f1e-71581c8d6a7e.png)

So overall, my question is: when I run gluTessBeginPolygon and pass in an array, how can I make it so the array I get back excludes the triangles generated for a particular contour?

**Comment by brendankenny** [comment]
Hi! I'm glad the library has helped. It was created for exactly the kind of tricky data you're talking about...basically libtess is always there as backup for those few cases earcut doesn't handle :)

Without seeing your data it's hard to know exactly what you need to do, but as a guess, you need to reverse the order of the vertices in one of your contours. libtess looks at every region covered by polygons and adds up a winding number for every contour over it. For counter-clockwise contours, it adds 1 to the winding number, for clockwise contours, it subtracts one. It then includes or excludes regions based on if the winding rule accepts or rejects the winding number for that region.

In your case, if you make the large triangle contour in counter-clockwise order and the circle contour in clockwise order, the outer area will have a winding number of 1, and the inner circular area will have a winding order of 1 - 1 = 0, so you can use GLU_TESS_WINDING_POSITIVE (which accepts only areas with winding numbers > 0) as the winding rule for excluding the inner circle (or, if your contours are the opposite way and the outer triangles are clockwise and the inner circle is counter-clockwise, you could use GLU_TESS_WINDING_NEGATIVE to get the same result).

Hopefully that's new information and I'm not repeating something you've already tried. Let me know if it works.

**Comment by iominh** [comment]
Thanks for the fast response! That information definitely was very helpful. I was able to mostly get my intended result of excluding the interior contour by playing around with vertex ordering and winding rules:

![tesselation_some_success](https://cloud.githubusercontent.com/assets/1434477/8086358/02ecd280-0f63-11e5-9fe1-70cbd2d13064.png)

I have a follow-up question if you don't mind: is there any way to exclude triangles not inside the original contours?

In a different example, here's the initial triangle contour (all colors below are randomized just to illustrate where the triangles are):
![test2_triangle](https://cloud.githubusercontent.com/assets/1434477/8086243/5b3e5784-0f62-11e5-845e-90efae51f777.jpg)

However, once I add a circle contour to exclude, it appears to return results outside of the original triangle contour (notice how the lower right half of New Mexico wasn't originally in the initial triangle):
![test1_issue](https://cloud.githubusercontent.com/assets/1434477/8086429/5eac60c2-0f63-11e5-93e4-e150a520c8ac.png)
More closeup screenshot of original triangle contour:
![screen shot 2015-06-10 at 11 35 46 am](https://cloud.githubusercontent.com/assets/1434477/8086729/e5ecb95a-0f64-11e5-9ce1-7968c5ed559d.png)
More closeup screenshot of original circle contour that I'm trying to exclude:
![circlecontour](https://cloud.githubusercontent.com/assets/1434477/8093902/c51ff412-0f91-11e5-8a7c-0b8f080d4050.png)
More closeup screenshot of tessellation results outside that triangle:
![screen shot 2015-06-10 at 11 33 47 am](https://cloud.githubusercontent.com/assets/1434477/8086697/baba4630-0f64-11e5-8b2b-3518f6b368fc.png)

I'm going to dive more into the documentation. I tried setting tessy.gluTessProperty(libtess.gluEnum.GLU_TESS_BOUNDARY_ONLY, true) but once I did that it didn't return any data back. I'll also try to come back with some more Codepens + sample data to more clearly illustrate what's going on.

**Comment by brendankenny** [comment]
Yeah, that's interesting. Geometry outside the original contours should never be generated, so something weird is going on. You can see that somewhat in the fact that the lower part of new mexico in the circle that shouldn't be covered at all appears to be covered by something new (though it's not clear from the screenshot by what exactly...doesn't look like a triangle). A codepen would be really helpful! That way I could play with it and see what's going on.

Feel free to use representative data (as long as it shows the same problem) if you don't want to share your current work publicly.

**Comment by brendankenny** [comment]
You may find the [original libtess README](http://cgit.freedesktop.org/mesa/glu/tree/src/libtess/README) helpful, though you'll have to mentally translate from the old C API to the mostly but not completely the same JS API. It's also not at all the clearest README, assuming familiarity with tessellators of the time (early 90s), but it's much more detailed than libtess.js's README :)

**Comment by iominh** [comment]
Hmmm, well, it looks like the triangle's top right and bottom left points are stretching slightly outside the original triangle contour to meet up with the circle contour's edges (I added another closeup screenshot of the triangulated circle contour above). As a quick fix I'm just going to use a [point-in-polygon](https://github.com/mikolalysenko/robust-point-in-polygon) algorithm to remove all triangleVerts that are either inside the original circle contour or outside the original triangle contour.

I also tossed up a github repo with sample data and code for this example (https://github.com/iominh/libtessTests/tree/master/test1). Hopefully it's helpful, it's always been kind of tricky figuring out how to visualize everything 

Also thanks once again for all your help, if there's any way I can help you back I'll be glad to do it (I went ahead and starred all your popular github repos lol)
- Debug update 6/10/15 6:45PM EDT
  I've identified one suspect triangle. It's the 23rd triangle returned from triangleVerts  [-103.33711242675781, 34.15195846557617, -135, 0, -90, 45]:
  ![23rdtriangleincorrect](https://cloud.githubusercontent.com/assets/1434477/8096040/f56a7ad4-0fa0-11e5-8b3e-edcac3851bca.png)
- Debug update 6/11/15 10:36AM EDT
  I ended up using https://github.com/Turfjs/turf-intersect, which was able to filter out the unwanted triangles within the circle contour.
